File: 3885-CountSpecialTriplets/3885-CountSpecialTriplets.py
# Last updated: 12/25/2025, 7:11:13 PM
class Solution:
    def specialTriplets(self, nums: List[int]) -> int:
        mod = 10**9 + 7

        prev = defaultdict(int)
        after = defaultdict(int)

        for num in nums:
            prev[num] += 1
        
        after[nums[-1]] += 1
        prev[nums[-1]] -= 1
        
        n = len(nums)
        ret = 0
        for i in range(n-2, 0, -1):
            prev[nums[i]] -= 1

            curr = nums[i] * 2
            if prev[curr] and after[curr]:
                ret += prev[curr] * after[curr] % mod
                ret %= mod
            
            after[nums[i]] += 1

        return ret
            
                
        

File: 3885-CountSpecialTriplets/3885-CountSpecialTriplets.cpp
// Last updated: 12/25/2025, 7:11:10 PM
const int N = 2e5 + 1;
int left_freq[N];
int right_freq[N];
int mod = 1e9 + 7;

class Solution {
public:
    int specialTriplets(vector<int>& nums) {
        int n = nums.size();

        memset(left_freq, 0, sizeof(left_freq));
        memset(right_freq, 0, sizeof(right_freq));

        for (int num: nums) {
            right_freq[num]++;
        }

        int ans = 0;
        for (int num: nums) {
            right_freq[num]--;

            int need = num * 2;
            ans += (left_freq[need] * 1LL * right_freq[need]) % mod;
            ans %= mod;

            left_freq[num]++;
        }

        return ans;
    }
};